test/src/test/java/jenkins/model/NodesTest.java

```diff
@@ -88,6 +90,28 @@ public void addNodeShouldReplaceExistingNode() throws Exception {
         assertThat(r.jenkins.getNode("foo"), sameInstance(newNode));
     }
 
+    @Test
+    @Issue("JENKINS-56403")
+    public void replaceNodeShouldRemoveOldNode() throws Exception {
+        Node oldNode = r.createSlave("foo", "", null);
+        Node newNode = r.createSlave("foo-new", "", null);
+        r.jenkins.addNode(oldNode);
+        r.jenkins.getNodesObject().replaceNode(oldNode, newNode);
+        r.jenkins.getNodesObject().load();
+        assertNull(r.jenkins.getNode("foo"));
+    }
+
+    @Test
+    @Issue("JENKINS-56403")
+    public void replaceNodeShouldNotRemoveIdenticalOldNode() throws Exception {
```

Comment:
This new test succeeds before and after the `core/src/main/java` change, but IIRC the goal is to harness more that side use case, even if this was not known to be an issue currently.